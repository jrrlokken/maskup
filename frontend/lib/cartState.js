import { createContext, useContext, useState } from 'react';

const LocalStateContext = createContext();
const LocalStateProvider = LocalStateContext.Provider;

const CartStateProvider = ({ children }) => {
  const [cartOpen, setCartOpen] = useState(false);
  const toggleCart = () => {
    setCartOpen(!cartOpen);
  }
  const closeCart = () => {
    setCartOpen(false);
  }
  const openCart = () => {
    setCartOpen(true);
  }

  return (
    <LocalStateProvider
      value={{
        cartOpen,
        setCartOpen,
        toggleCart,
        closeCart,
        openCart
      }}
    >
      {children}
    </LocalStateProvider>
  );
}

const useCart = () => {
  const all = useContext(LocalStateContext);
  return all;
}

export { CartStateProvider, useCart };