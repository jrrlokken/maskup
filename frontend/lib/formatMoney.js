export default function formatMoney(amount) {
  const options = {
    style: 'currency',
    currency: 'USD',
    minimumFractionDigits: 2
  };
  const formatter = Intl.NumberFormat('en-US', options);
  return formatter.format(amount / 100);
}