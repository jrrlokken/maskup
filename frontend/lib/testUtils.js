import casual from 'casual';

// seed it so we get consistent results
casual.seed(777);

const type = (wrapper, name, value) => {
  wrapper.find(`input[name='${name}']`).simulate('change', {
    target: { name, value }
  });
};

const fakeProduct = ({
  // __typename: 'Item',
  id: 'abc123',
  name: 'Test Product',
  description: 'This is a test product',
  price: 1499,
  photo: {
    altText: 'dog',
    image: {
      publicUrlTransformed: 'dog.jpg',
    },
  },
  user: null,
});

const fakeUser = overrides => ({
  // __typename: 'User',
  id: '4234',
  name: casual.name,
  email: casual.email,
  permissions: ['ADMIN'],
  orders: [],
  cart: [],
  ...overrides,
});

const fakeOrderItem = () => ({
  // __typename: 'OrderItem',
  id: casual.uuid,
  photo: {
    image: {
      publicUrlTransformed: `${casual.word}.jpg`
    }
  },
  name: casual.words(),
  price: 4234,
  quantity: 1,
  description: casual.words(),
});

const fakeOrder = () => ({
  // __typename: 'Order',
  id: 'ord123',
  charge: 'ch_123',
  total: 40000,
  items: [fakeOrderItem(), fakeOrderItem()],
  createdAt: '2022-12-11T20:16:13.797Z',
  user: fakeUser(),
});

const fakeCartItem = overrides => ({
  // __typename: 'CartItem',
  id: 'omg123',
  quantity: 3,
  product: fakeProduct,
  user: fakeUser(),
  ...overrides,
});


// Fake LocalStorage
class LocalStorageMock {
  constructor() {
    this.store = {};
  }

  clear() {
    this.store = {};
  }

  getItem(key) {
    return this.store[key] || null;
  }

  setItem(key, value) {
    this.store[key] = value.toString();
  }

  removeItem(key) {
    delete this.store[key];
  }
}

export {
  type,
  fakeProduct,
  fakeUser,
  fakeCartItem,
  fakeOrder,
  fakeOrderItem,
  LocalStorageMock,
};