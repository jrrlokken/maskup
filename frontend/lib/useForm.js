import { useState, useEffect } from 'react';

const useForm = (initial = {}) => {
  const [inputs, setInputs] = useState(initial);
  const initialValues = Object.values(initial).join('');

  useEffect(() => {
    setInputs(initial);
  }, [initialValues]);

  const handleChange = event => {
    let { value, name, type } = event.target;
    if (type === 'number') {
      value = parseInt(value);
    }
    if (type === 'file') {
      [value] = event.target.files;
    }
    setInputs({
      ...inputs,
      [name]: value,
    });
  }

  const resetForm = () => {
    setInputs(initial);
  }

  const clearForm = () => {
    const blankState = Object.fromEntries(
      Object.entries(inputs).map(([key]) => [key, ''])
    );
    setInputs(blankState);
  }

  return {
    inputs,
    handleChange,
    resetForm,
    clearForm
  };
}

export default useForm;