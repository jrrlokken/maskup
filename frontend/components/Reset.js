import { useMutation } from '@apollo/client';
import gql from 'graphql-tag';
import { useRouter } from 'next/router';

import useForm from '../lib/useForm';
import Form from './styles/Form';
import DisplayError from './ErrorMessage';

const RESET_MUTATION = gql`
  mutation RESET_MUTATION(
    $token: String!
    $password: String!
    $email: String!
  ) {
    redeemUserPasswordResetToken(
      token: $token
      password: $password
      email: $email
    ) {
      code
      message
    }
  }
`;

const Reset = ({ token }) => {
  const { inputs, handleChange, resetForm } = useForm({
    password: '',
    email: '',
    token,
  });
  const [reset, { data, error, loading }] = useMutation(
    RESET_MUTATION, {
      variables: inputs,
  });
  const router = useRouter();
  const successfulError = data?.redeemUserPasswordResetToken?.code
    ? data?.redeemUserPasswordResetToken
    : undefined;
  async function handleSubmit(event) {
    event.preventDefault();
    await reset().catch(console.error);
    resetForm();
    router.push('/signin');
  }

  return (
    <Form
      method='POST'
      onSubmit={handleSubmit}
    >
      <h2>Reset your password</h2>
      <DisplayError error={error || successfulError} />
      <fieldset>
        {data?.redeemUserPasswordResetToken === null && (
          <p>Success! You can now sign in.</p>
        )}
        <label htmlFor='email'>
          Email address
          <input
            type='email'
            name='email'
            placeholder='Your Email Address'
            value={inputs.email}
            onChange={handleChange}
            required
          />
        </label>
        <label htmlFor='password'>
          Password
          <input
            type='password'
            name='password'
            placeholder='Password'
            value={inputs.password}
            onChange={handleChange}
            required
          />
        </label>
        <button type='submit'>Reset</button>
      </fieldset>
    </Form>
  );
}

export default Reset;