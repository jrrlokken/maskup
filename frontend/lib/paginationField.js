import { PAGINATION_QUERY } from '../components/Pagination';

const paginationField = () => {
  return {
    keyArgs: false,
    read(existing = [], { args, cache }) {
      const { skip, first } = args;
      const data = cache.readQuery({ query: PAGINATION_QUERY });
      const count = data?._allProductsMeta?.count;
      const page = skip / first + 1;
      const pages = Math.ceil(count / first);

      const items = existing
        .slice(skip, skip + first)
        .filter((x) => x);

      if (items.length && items.length !== first && page === pages) {
        return items;
      }
      if (items.length !== first) {
        return;
      }
      if (items.length) {
        return items;
      }
      return false;
    },

    merge(existing, incoming, { args }) {
      const { skip, first } = args;
      const merged = existing ? existing.slice(0) : [];
      for (let i = skip; i < skip + incoming.length; ++i) {
        merged[i] = incoming[i - skip];
      }
      return merged;
    },
  };
}

export default paginationField;