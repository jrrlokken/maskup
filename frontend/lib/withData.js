import { ApolloClient, ApolloLink, InMemoryCache } from '@apollo/client';
import { onError } from '@apollo/link-error';
import { getDataFromTree } from '@apollo/react-ssr';
import { createUploadLink } from 'apollo-upload-client';
import withApollo from 'next-with-apollo';

import { endpoint, productionEndpoint } from '../config';
import paginationField from './paginationField';

const createClient = ({ headers, initialState }) => {
  return new ApolloClient({
    link: ApolloLink.from([
      onError(({ graphQLErrors, networkError }) => {
        if (graphQLErrors)
          graphQLErrors.forEach(({ message, locations, path }) =>
            console.log(
              `[GraphQL error]: Message: ${message}, Location: ${locations}, Path: ${path}`
            )
          );
        if (networkError)
          console.log(
            `[Network error]: ${networkError}. Backend is unreachable. Is it running?`
          );
      }),

      createUploadLink({
        uri: process.env.NODE_ENV === 'development' ? endpoint : productionEndpoint,
        fetchOptions: {
          credentials: 'include',
        },
        // pass the headers along from this request. This enables SSR with logged in state
        headers,
      }),
    ]),
    cache: new InMemoryCache({
      typePolicies: {
        Query: {
          fields: {
            allProducts: paginationField(),
          },
        },
      },
    }).restore(initialState || {}),
  });
}

export default withApollo(createClient, { getDataFromTree });
