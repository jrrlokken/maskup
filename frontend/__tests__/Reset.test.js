import { configure, mount } from 'enzyme';
import Adapter from '@hteker/enzyme-adapter-react-17';
import toJSON from 'enzyme-to-json';
import { act } from 'react-dom/test-utils';
import wait from 'waait';
import { MockedProvider } from '@apollo/client/testing';
import { fakeProduct } from '../lib/testUtils';

import Reset, { RESET_MUTATION } from '../components/Reset';

configure({ adapter: new Adapter() });

const token = 'asdfjaskdlasdflkjasldf';

const mocks = [
  {
    request: { 
      query: RESET_MUTATION,
      variables: {
        redeemUserPasswordResetToken: {
          token,
          email: 'joshualokken@pm.me',
          password: 'password'
        }
      }
    },
    result: {
      code: undefined,
      message: 'Message'
    }
  }
];

describe('<Reset/>', () => {
  it('renders and matches snapshot', () => {
    const wrapper = mount(
      <MockedProvider>
        <Reset />
      </MockedProvider>
    );
    const form = wrapper.find('form');
    expect(toJSON(form)).toMatchSnapshot();
    expect(form.find('h2').text()).toBe('Reset your password');
    expect(form.find('button').text()).toBe('Reset');
  });
  it('calls the mutation', async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks} addTypename={false}>
        <Reset />
      </MockedProvider>
    );
    wrapper.find('button').simulate('click');
    await act(async () => {
      await wait();
      wrapper.update();
    });
    console.log(wrapper.debug());
  })
});
