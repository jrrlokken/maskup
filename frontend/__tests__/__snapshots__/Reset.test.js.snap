// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Reset/> renders and matches snapshot 1`] = `
<form
  className="Form-c7byzo-0 csfyM"
  method="POST"
  onSubmit={[Function]}
>
  <h2>
    Reset your password
  </h2>
  <DisplayError
    error={Object {}}
  />
  <fieldset>
    <label
      htmlFor="email"
    >
      Email address
      <input
        name="email"
        onChange={[Function]}
        placeholder="Your Email Address"
        required={true}
        type="email"
        value=""
      />
    </label>
    <label
      htmlFor="password"
    >
      Password
      <input
        name="password"
        onChange={[Function]}
        placeholder="Password"
        required={true}
        type="password"
        value=""
      />
    </label>
    <button
      type="submit"
    >
      Reset
    </button>
  </fieldset>
</form>
`;
